Spring Boot Portfolio Project — Part 7. JWT. Executor: Oleksii Bashkirov.

OleksiiBashkirov opened the pull request:
feat: Implement JWT authentication

- Added `login(UserLoginRequestDto request)` method to `AuthenticationController`.
- Created `AuthenticationService`, `JwtAuthenticationFilter`, and `JwtUtil` classes.
- Made `/auth/login` endpoint publicly accessible.
- Configured `SecurityFilterChain` to use the new JWT filter.
- Implemented data validation for new DTO classes.

OleksiiBashkirov committed:
Fixed security configuration and liquibase issue

OleksiiBashkirov committed:
Fixed checkstyle errors

OleksiiBashkirov committed:
- Removed the redundant Repository annotation in the RoleRepository interface.

OleksiiBashkirov committed:
- Updated pom.xml file;
- Added annotations in SecurityConfig;
- Included userDetailsService in SecurityConfig;
- Removed EAGER fetch type for roles in User entity;
- Updated CustomUserDetailsService to use EntityNotFoundException;
- Implemented role assignment in UserServiceImpl during user registration.

OleksiiBashkirov committed:
- Removed debug messages.

OleksiiBashkirov committed:
1. Updated SecurityConfig.java:
   - Removed the @EnableWebSecurity annotation.

2. Updated BookMapper.java:
   - Removed the @Component annotation.

3. Updated User.java:
   - Corrected @ManyToMany annotation formatting.

4. Updated CustomUserDetailsService.java:
   - Refactored the "loadUserByUsername" method.

5. Updated UserServiceImpl.java:
   - Replaced Collections.singleton with Set.of for setting user roles.

OleksiiBashkirov committed:
feat: Implement JWT authentication

- Added `login(UserLoginRequestDto request)` method to `AuthenticationController`.
- Created `AuthenticationService`, `JwtAuthenticationFilter`, and `JwtUtil` classes.
- Made `/auth/login` endpoint publicly accessible.
- Configured `SecurityFilterChain` to use the new JWT filter.
- Implemented data validation for new DTO classes.

Elena-Bruyako commented on pom.xml:
```diff
@@ -101,6 +102,31 @@
+            <version>${jjwt.version}</version>
+            <scope>runtime</scope>
+        </dependency>
+
```
remove empty lines

Elena-Bruyako commented on src/main/java/mate/academy/config/JwtConfig.java:
```diff
@@ -0,0 +1,20 @@
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class JwtConfig {
```
Why do you need this?

Elena-Bruyako commented on src/main/java/mate/academy/config/SecurityConfig.java:
```diff
@@ -0,0 +1,94 @@
+        return new BCryptPasswordEncoder();
+    }
+
+    @Bean
```
This confi could be much simplier
```suggestion
    @Bean
    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
        return http
                .cors(AbstractHttpConfigurer::disable)
                .csrf(AbstractHttpConfigurer::disable)
                .authorizeHttpRequests(
                        auth -> auth
                                .requestMatchers(
                                        antMatcher("/auth/**"),
                                        antMatcher("/swagger-ui/**"),
                                        antMatcher("/v3/api-docs/**"))
                                .permitAll()
                                .anyRequest()
                                .authenticated()
                )
                .sessionManagement(
                        s -> s.sessionCreationPolicy(SessionCreationPolicy.STATELESS))
                .addFilterBefore(jwtAuthFilter,
                        UsernamePasswordAuthenticationFilter.class)
                .userDetailsService(service)
                .build();
    }

    @Bean
    public AuthenticationManager authenticationManager(
            AuthenticationConfiguration authenticationConfiguration) throws Exception {
        return authenticationConfiguration.getAuthenticationManager();
    }
```

Elena-Bruyako commented on src/main/java/mate/academy/config/SwaggerConfig.java:
```diff
@@ -0,0 +1,23 @@
+    @Bean
+    public OpenAPI customOpenApi() {
+        return new OpenAPI()
+                .components(new Components().addSecuritySchemes("BearerAuth",
```
create constant for "BearerAuth"

Elena-Bruyako commented on src/main/java/mate/academy/dto/UserLoginResponseDto.java:
```diff
@@ -0,0 +1,10 @@
+import lombok.Data;
+
+@Data
+@AllArgsConstructor
```
```suggestion
```

Elena-Bruyako commented on src/main/java/mate/academy/repository/UserRepository.java:
```diff
@@ -1,10 +1,11 @@
 
-@Repository
 public interface UserRepository extends JpaRepository<User, Long> {
+    Optional<User> findByEmail(String email);
```
```suggestion
    @EntityGraph(attributePaths = "roles")
    Optional<User> findByEmail(String email);
```

Elena-Bruyako commented on src/main/java/mate/academy/security/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    private String getToken(HttpServletRequest httpServletRequest) {
+        String bearerToken = httpServletRequest.getHeader("Authorization");
```
```suggestion
        String bearerToken = httpServletRequest.getHeader(HttpHeaders.AUTHORIZATION);
```

Elena-Bruyako commented on src/main/java/mate/academy/security/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,62 @@
+
+    private String getToken(HttpServletRequest httpServletRequest) {
+        String bearerToken = httpServletRequest.getHeader("Authorization");
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
```
Create constant for "Bearer "

Elena-Bruyako commented on src/main/java/mate/academy/security/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,62 @@
+    private String getToken(HttpServletRequest httpServletRequest) {
+        String bearerToken = httpServletRequest.getHeader("Authorization");
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
+            return bearerToken.substring(7);
```
```suggestion
            return bearerToken.substring(constant.length());
```

Elena-Bruyako reviewed (changes requested):
see comments

OleksiiBashkirov committed:
Implement suggestions and improvements

- Removed empty lines in pom.xml;
- Removed JwtConfig.java;
- Simplified SecurityConfig.java by adjusting the security filter chain and configurations;
- Added constants for "BearerAuth" and "Bearer " in appropriate classes;
- Updated UserLoginResponseDto;
- Enhanced UserRepository with @EntityGraph for roles in findByEmail method;
- Replaced hardcoded "Authorization" header string with HttpHeaders.AUTHORIZATION constant;
- Ensured proper substring usage in JwtAuthenticationFilter.

OleksiiBashkirov commented:
I wanted to apologize for my recent accidental pull request to https://github.com/OleksiiBashkirov/online-book-store/pull/8. 
Due to an accidental click, I submitted the changes meant for the "JWT" theme under the "Spring Boot Security" theme.
Sorry about that. Please disregerd that pull request.
Thank you for your understanding and patience!

OleksiiBashkirov committed:
- Deleted unnecessary class `JwtConfig`;
- Refactored constant `BEARER_AUTH` in `SwaggerConfig`;

MentorMateAcademy commented on src/main/java/mate/academy/config/SwaggerConfig.java:
```diff
@@ -0,0 +1,24 @@
+                        new SecurityScheme()
+                                .type(SecurityScheme.Type.HTTP)
+                                .scheme("bearer")
+                                .bearerFormat("JWT")))
```
you can save to constant too

OleksiiBashkirov committed:
- Added constants for "bearer" and "JWT" in SwaggerConfig class;

OleksiiBashkirov committed:
- Fixed checkstyle errors;

Elena-Bruyako commented on src/main/java/mate/academy/controller/AuthenticationController.java:
```diff
@@ -16,10 +19,16 @@
         return userService.register(requestDto);
     }
+
+    @PostMapping("/login")
```
Add swagger annotations

Elena-Bruyako reviewed (changes requested):
This branch has conflicts, please fix them

OleksiiBashkirov committed:
Merge branch 'main' into jwt-hw
